python/nano/src/bigdl/nano/tf/hpo/tunemixin.py

```diff
@@ -0,0 +1,94 @@
+#
+# Copyright 2016 The BigDL Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from .objective import Objective
+import optuna
+
+
+class HPOMixin:
+
+    def tune(
+        self,
+        n_trails=1,
+        resume=False,
+        target_metric="accuracy",
+        direction="maximize",
+        timeout=600,
+        *args,
+        **kwargs
+    ):
+        """ Do the hyper param tuning.
+
+        Args:
+            n_trails (int, optional): number of trials to run. Defaults to 1.
+            resume (bool, optional): whether to resume the previous tuning. Defaults to False.
+            target_metric (str, optional): the target metric to optimize. Defaults to "accuracy".
+            direction (str, optional): optimize direction. Defaults to "maximize".
+            timeout (int, optional): optuna argument. Defaults to 600.
+        """
+        if not resume:
+            self.objective = Objective(
+                model_cls=self.__class__.__bases__[1],
+                model_initor=self.model_initor,
+                model_compiler=self.model_compiler,
+                target_metric=target_metric,
+                *args,
+                **kwargs,
+            )
+            self.study = optuna.create_study(direction=direction)
+        self.study.optimize(self.objective, n_trials=n_trails, timeout=timeout)
+
+    def tune_summary(self):
+        """Retrive a summary of trials
+
+        Returns:
+            dataframe: A summary of all the trials
+        """
+        if self.study is not None:
+            print("Number of finished trials: {}".format(len(self.study.trials)))
+            best = self.study.best_trial
+            print("Best trial:")
+            print("  Value: {}".format(best.value))
+            print("  Params: ")
+            for key, value in best.params.items():
+                print("    {}: {}".format(key, value))
+            return self.study.trials_dataframe(attrs=("number", "value", "params", "state"))
+        else:
+            print("Seems you have not done any tuning yet.  \
+                  Call tune and then call tune_summary to get the statistics.")
+
+    def end_tune(self, use_trial_id=-1):
```

Comment:
how does the user specify use_trial_id?

Reply:
user can call model.tune_summary() to get a dataframe that contains all trial results. Each row in the data frame is one trail. Columns include trial_id, value of the hyperparmeters and the evaluation result. If user_trial_id is not set, the trial with the best evaluation results will be used automatically. Or user can select the id he wants to use and pass it in use_trial_id.